### ISSUE TITLE
Break optimization triggering other optimization in canonicalize

### ISSUE THREAD

**Issue by nouiz** [Theano/Theano#4912]
In the canonicalize Equilibrium, we want to lower the number of iteration to speed up the compilation time.

For this, we need each optimization to return as much as possible to return a pre optimized graph. Told in a different way, we need each optimization to not rely on other optimization to clean up what it does.

For this, we need to know when an optimization create a new nodes, does another optimization apply. If so, could we have generated a better graph at first?

I see a few way to find which optimization is triggered by which optimization:

1) check the EQ profile output. See if you can infer which optimization trigger which one. More possible in the last few iteration of the EQ.
2) In the EQ, after the first pass, print which opt apply on which node, but also print node.tag.imported_by. That containt the optimization that imported that node in the graph.
3) This is a follow up of 2), mostly we can the information from 2) in a dict with a key (current opt, node.tag.imported_by) and the value is an int that we increment. Then we can sort on the value the dict and print it in the print_profile method. This will allow to see which sequence of optimization happen frequently.

**Comment by Sentient07** [comment]
I did the second and 3rd part. I have filtered out only the top 10 opts and node created. Also, i don't understand why I see `init` in `node.tag.imported_by`.

**Comment by nouiz** [comment]
"init" mean that this node was there in the original graph. It was not
created by other optimization.

**Comment by nouiz** [comment]
Can you make a PR with this code and put in it the result you have?

**Comment by Sentient07** [comment]
Yeah, the printing that's happening is still a lot and clumsy and I'm finding myself difficult to understand. I'll make it readable before opening

**Comment by Sentient07** [comment]
I'm trying to fix an indentation in the printing of profile from print_profile. But this is what i get from printing inside the apply after the first pass, http://pastebin.com/ggfXkHLv

**Comment by nouiz** [comment]
You need to put that in a dict and count the number of time each opt work
on variable
key=(imported_by, local_opt)
value= number of time this combination appened

On Fri, Sep 2, 2016 at 3:27 PM, Ramana Subramanyam <notifications@github.com

**Comment by Sentient07** [comment]
That is already done by `process_count` right ?

**Comment by Sentient07** [comment]
Oh, you mean to say the combination of {opt: node_created} 's count ?

**Comment by nouiz** [comment]
no. The combination of imported_by and opt.

Mostly, what you print on 2 lines. Make a tuple of that. Each time that
this tuple appear, count this as 1 event. I want the count of each event.

On Fri, Sep 2, 2016 at 3:37 PM, Ramana Subramanyam <notifications@github.com

**Comment by Sentient07** [comment]
I have a quick question. If i have they key as (imported, opt) then what we count won't be accurate because i'm using only the first ten node.tag.imported_by but all the opt used. How to bridge this inconsistency ?

**Comment by Sentient07** [comment]
I've added that dict pair and this is the profiling result currently. http://pastebin.com/pqkGKAtK 
In this , i've set the `ps.fct_callcount` to >1 inside the profiling.py

**Comment by Sentient07** [comment]
From the profiling, we see that 
```
     Opts Used : local_useless_fill
     Nodes imported : ('local_mul_zero',)
      Number of pairs found : 617
```
This isn't abnormal, I suppose ? From what i guess, if some operation is being done with  a zero tensor, then instead of performing the multiplication, we're replacing. This seems normal that there are these many pairs as it is common to have those many operations.

**Comment by Sentient07** [comment]
> For this, we need to know when an optimization create a new nodes, does another optimization apply. If so, could we have generated a better graph at first?

If I'm understanding this correctly, you mean the new node created by an optimizer being optimized by one more optimizer is redundant and reducing that is the major task, right ?

**Comment by Sentient07** [comment]
The most frequent node introduced by an optimizer and the optimizer that introduces the node : http://pastebin.com/qSRrzm1s -> This is with optimizer=fast_compile
http://pastebin.com/FW5G8Hc9 -> This is with optimzier=fast_run.

P.S : There is a small grammatical error. It isn't the node imported, it is which optimizer that imports the node.

**Comment by Sentient07** [comment]
@nouiz In case of a Global Optimizer introducing a new node, I need to be more specific with what local_optimizer inside the GlobalOpt introduced ?
